hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1003 @@
+    }
+
+    /*
+    // TODO: We may not be able to sync in certain cases (instants archived etc)
+    //if (!canSync(datasetMetaClient)) {
+      // Need to recreate the table as sync has failed
+      // TODO: delete the table
+    //  initFromFilesystem(datasetMetaClient);
+    //}
+    */
+
+    // This is always called even in case the table was created for the first time. This is because
+    // initFromFilesystem() does file listing and hence may take a long time during which some new updates
+    // may have occurred on the table. Hence, calling this always ensures that the metadata is brought in sync
+    // with the active timeline.
+    syncFromInstants(datasetMetaClient);
+
+    // Publish some metrics
+    if (context != null) {
+      long durationInMs = metrics.getDurationInMs(context.stop());
+      // Time to initilize and sync
+      if (exists) {
+        metrics.updateMetadataMetrics(INITIALIZE_STR, 0, 0);
+        metrics.updateMetadataMetrics(SYNC_STR, durationInMs, 1);
+      } else {
+        metrics.updateMetadataMetrics(INITIALIZE_STR, durationInMs, 1);
+        metrics.updateMetadataMetrics(SYNC_STR, 0, 0);
+      }
+
+      // Total size of the metadata and count of base/log files
+      long totalBaseFileSizeInBytes = 0;
+      long totalLogFileSizeInBytes = 0;
+      int baseFileCount = 0;
+      int logFileCount = 0;
+      HoodieTableFileSystemView fsView = new HoodieTableFileSystemView(metaClient, metaClient.getActiveTimeline());
+      List<FileSlice> latestSlices = fsView.getLatestFileSlices(METADATA_PARTITION_NAME).collect(Collectors.toList());
+
+      for (FileSlice slice : latestSlices) {
+        if (slice.getBaseFile().isPresent()) {
+          totalBaseFileSizeInBytes += fs.getFileStatus(new Path(slice.getBaseFile().get().getPath())).getLen();
+          ++baseFileCount;
+        }
+        Iterator<HoodieLogFile> it = slice.getLogFiles().iterator();
+        while (it.hasNext()) {
+          totalLogFileSizeInBytes += fs.getFileStatus(it.next().getPath()).getLen();
+          ++logFileCount;
+        }
+      }
+
+      metrics.updateMetadataSizeMetrics(totalBaseFileSizeInBytes, totalLogFileSizeInBytes, baseFileCount, logFileCount);
+    }
+  }
+
+  /**
+   * Initialize the Metadata Table by listing files and partitions from the file system.
+   *
+   * @param datasetMetaClient {@code HoodieTableMetaClient} for the dataset
+   */
+  private void initFromFilesystem(HoodieTableMetaClient datasetMetaClient) throws IOException {
+    ValidationUtils.checkState(!readOnly, "Metadata table cannot be initialized in readonly mode");
+
+    // If there is no commit on the dataset yet, use the SOLO_COMMIT_TIMESTAMP as the instant time for initial commit
+    Option<HoodieInstant> latestInstant = datasetMetaClient.getActiveTimeline().filterCompletedInstants().lastInstant();
+    String createInstantTime = latestInstant.isPresent() ? latestInstant.get().getTimestamp() : SOLO_COMMIT_TIMESTAMP;
```

Comment:
Seems like you will use this timestamp for the intial base file. This is same as HoodieTimeline.INIT_INSTANT_TS. Can we reuse that constant? and more importantly, would this cause any issues when we use the bootstrap feature? 
In this scenario, there will be a 0000000000 commit on the dataset timeline. and you will attempt to also create the same instant first during initing and then subsequently for the actual write?